### ISSUE TITLE
Docs for TMEM type RPC parameters?

### ISSUE THREAD

**Issue by lws-team** [OP-TEE/optee_os#1283]
I am able to send value type RPC parameters around OK.

However when trying to use TMEM type RPC, I get an abort when performing the RPC itself.

```
[  142.374973] pgd = ffff80003bbc7000
[  142.378490] [300000000] *pgd=000000003a8f1003
```

The idea is send three params, one value, one TMEM going to tee-supplicant and one TMEM coming back.  I marshal the complex parameters going in both directions inside the TMEM blocks.

At the moment, I attempt to reuse tee_fs_rpc_cache_alloc() to allocate the pa and get the va, as I saw it was fone in the fs code.  But I found using it twice, once for each TMEM gave back the same va both times.  So I tried allocating one chunk big enough for both... it's still blowing chunks.

Googling around for advice on how TMEM RPC is supposed to be used, I didn't find anything.

Any suggestions?

**Comment by lws-team** [comment]
I stripped the tee_fs_rpc_cache_alloc() api I was abusing down to the relevant pieces and used them directly.  And I shifted to using one INOUT tmem block for both parameters.

Then I noticed the console logging is disordered and partially hidden by the two worlds logging to the same device... the code was actually dying in my RPC handler in tee-supplicant, much later than I thought.

Finally by moving around an early failed return from my RPC handler, rather than trying to "debug by printf" when the console output was incomplete, I was able to spot the segfault was completely my own fault (literally) dereferencing an uninitialized local pointer in a log message.

I'll leave this open a little longer since the whole end-end flow doesn't work yet and there may be related questions.

**Comment by lws-team** [comment]
... is there a restriction about dynamically loaded TAs being able to use malloc()'d data from the core during a syscall?

Eg... dynamic TAs don't map the whole malloc pool region for security reasons?  How should we map the malloc'd region if so?

**Comment by lws-team** [comment]
I arranged to do the malloc() on the TA side, where it would certainly be mapped in the TA and presumably the core code VM table must be able to see the whole malloc pool at once: subsequently I fixed the cause of either the same or another data abort, that was my fault again.

Since that left me in a state of doubt about malloc, after it was working, in order to try to learn if this mapping issue was real or not, I patched the working code to do the malloc on the core side again.  That got me

```
ERROR:   TEE-CORE: user TA data-abort at address 0x3f058c74 (read permission fault)
```

after execution had returned through the syscall back into the dynamic TA.

It really looks like a dynamic TA cannot use a malloc'd buffer generated in core code via a syscall, because it's not mapped.  Can someone confirm it?  And if there is an api to map the core-malloc'd buffer into the dynamic TA if so?

**Comment by jforissier** [comment]
@lws-team 

> It really looks like a dynamic TA cannot use a malloc'd buffer generated in core code via a syscall, because it's not mapped. Can someone confirm it?

Correct. The opposite would be pretty unsafe/unsecure, right?

> And if there is an api to map the core-malloc'd buffer into the dynamic TA if so?

No. You have to manipulate some tables by yourself, and that's not exactly easy. See https://github.com/OP-TEE/optee_os/blob/master/core/arch/arm/mm/tee_mmu.c, functions `tee_mmu_map_stack()`, `tee_mmu_map_add_segment()`, `tee_mmu_map_param()`.

**Comment by lws-team** [comment]
> Correct. The opposite would be pretty unsafe/unsecure, right?

Yes I could understand why it might be so.  But I didn't see it told anywhere that it **is** so... generally OP-TEE models itself on the kernel, which is good, but this is a departure; kernel modules have no such mapping discontiguity.  (For good reasons in OP-TEE context...).

> No. You have to manipulate some tables by yourself, and that's not exactly easy. See https://github.com/OP-TEE/optee_os/blob/master/core/arch/arm/mm/tee_mmu.c, functions tee_mmu_map_stack(), tee_mmu_map_add_segment(), tee_mmu_map_param().

Thanks... I did the most messy api on my list second, getaddrinfo() (first was socket, that could work end-end using value references at all).  Actually getaddrinfo most naturally wants to have the core code side do the malloc.  But it was not too difficult for that set of parameters to have the malloc done in the TA-side wrapper.

I'll keep going and see if there are any cases I meet that really really want to have an official API for adopting core malloc()s into TA space.  Or if it can be avoided I'll avoid it.

Basically anyone who wants blocking semantics in their dynamic TA is into this flow of providing syscall wrapping so it can be used in a dynamic TA at all, and then the RPC stuff that AIUI must be done in core.  Then allocations made in core break back in the dynamic TA.  So it seems this should be documented somewhere if not already.
